stream_batch.go

```diff
@@ -0,0 +1,220 @@
+package scylla_cdc
+
+import (
+	"context"
+	"encoding/base64"
+	"sync/atomic"
+	"time"
+
+	"github.com/gocql/gocql"
+)
+
+type streamBatchReader struct {
+	config         *ReaderConfig
+	generationTime time.Time
+	streams        []StreamID
+	keyspaceName   string
+	tableName      string
+
+	lastTimestamp gocql.UUID
+	endTimestamp  atomic.Value
+
+	interruptCh chan struct{}
+}
+
+func newStreamBatchReader(
+	config *ReaderConfig,
+	generationTime time.Time,
+	streams []StreamID,
+	keyspaceName string,
+	tableName string,
+	startFrom gocql.UUID,
+) *streamBatchReader {
+	return &streamBatchReader{
+		config:         config,
+		generationTime: generationTime,
+		streams:        streams,
+		keyspaceName:   keyspaceName,
+		tableName:      tableName,
+
+		lastTimestamp: startFrom,
+
+		interruptCh: make(chan struct{}, 1),
+	}
+}
+
+func (sbr *streamBatchReader) run(ctx context.Context) error {
+	baseTableName := sbr.keyspaceName + "." + sbr.tableName
+
+	oldProgress := make(map[string]gocql.UUID, len(sbr.streams))
+	for _, stream := range sbr.streams {
+		progress, err := sbr.config.ProgressManager.GetProgress(sbr.generationTime, baseTableName, stream)
+		if err != nil {
+			return err
+		}
+		oldProgress[string(stream)] = progress.Time
+	}
+
+	minTime := oldProgress[string(sbr.streams[0])]
+	for _, progress := range oldProgress {
+		if compareTimeuuid(minTime, progress) > 0 {
+			minTime = progress
+		}
+	}
+	if (minTime != gocql.UUID{}) {
+		sbr.lastTimestamp = minTime
+	}
+
+	sbr.config.Logger.Printf("running batch starting from %v", sbr.lastTimestamp.Time())
+
+	input := CreateChangeConsumerInput{
+		TableName: baseTableName,
+		StreamIDs: sbr.streams,
+	}
+	consumer, err := sbr.config.ChangeConsumerFactory.CreateChangeConsumer(input)
+	if err != nil {
+		sbr.config.Logger.Printf("error while creating change consumer (will quit): %s", err)
+	} else {
+		defer consumer.End()
+	}
+
+	crq := newChangeRowQuerier(sbr.config.Session, sbr.streams, sbr.keyspaceName, sbr.tableName)
+
+	// sbr.config.Logger.Printf("starting stream processor loop for %v", sbr.streams)
+outer:
+	for {
+		timeWindowEnd := sbr.lastTimestamp.Time().Add(sbr.config.Advanced.QueryTimeWindowSize)
+		confidenceWindowEnd := time.Now().Add(-sbr.config.Advanced.ConfidenceWindowSize)
+
+		if timeWindowEnd.After(confidenceWindowEnd) {
+			timeWindowEnd = confidenceWindowEnd
+		}
+
+		pollEnd := gocql.MinTimeUUID(timeWindowEnd)
+
+		var (
+			err      error
+			rowCount int
+		)
+
+		readUpTo := pollEnd
+
+		if compareTimeuuid(sbr.lastTimestamp, pollEnd) < 0 {
+			// Set the time interval from which we need to return data
+			var iter *changeRowIterator
+			iter, err = crq.queryRange(sbr.lastTimestamp, pollEnd)
```

Comment:
`queryRange(x1, x2)` reads the open range `(x1, x2)`
this loop will thus be querying ranges `(x1, x2)`, `(x2, x3)`, `(x3, x4)`, etc.
this means you will be missing changes whose timeuuids are equal to these range bounds (`x1, x2, x3, x4`, ...), right?
Unless we assume that those timeuuids are never equal to `MinTimeUUID(someTime)`, which is probably valid, but I don't know

Reply:
I changed it so that now the range is closed on the right side - `(x1, x2]`. Let's not risk some row having a timeuuid equal to `MinTimeUUID`.